### PR TITLE
Add latest cwltool requirements

### PR DESCRIPTION
**Pull request by david4096**
**Description**

#5921 showed that the latest Python (3.12.x) is not compatible with previous versions of cwltool. This PR brings the cwltool requirements.txt offered by dockstore into the latest.

```
➜  dockstore git:(feature/dock-2539/cwltool-deps) ✗ curl -so requirements.txt "http://localhost:8080/metadata/runner_dependencies?client_version=1.15.3&python_version=3"
➜  dockstore git:(feature/dock-2539/cwltool-deps) ✗ python3 -m venv env
➜  dockstore git:(feature/dock-2539/cwltool-deps) ✗ python3 --version
Python 3.12.4
➜  dockstore git:(feature/dock-2539/cwltool-deps) ✗ source env/bin/activate
(env) ➜  dockstore git:(feature/dock-2539/cwltool-deps) ✗ pip install -r requirements.txt 

........

(env) ➜  dockstore git:(feature/dock-2539/cwltool-deps) ✗ cwltool --version
/Users/davidcs/git/dockstore/env/bin/cwltool 3.1.20240708091337
```

**Review Instructions**

Ensure cwltool can be installed from the requirements provided via the API using the above method

**Issue**

#5921

**Security and Privacy**

If there are any concerns that require extra attention from the security team, highlight them here and check the box when complete. 

- [x] Security and Privacy assessed

e.g. Does this change...
* Any user data we collect, or data location?
* Access control, authentication or authorization?
* Encryption features?

Please make sure that you've checked the following before submitting your pull request. Thanks!

- [x] Check that you pass the basic style checks and unit tests by running `mvn clean install`
- [x] Ensure that the PR targets the correct branch. Check the milestone or fix version of the ticket.
- [x] Follow the existing JPA patterns for queries, using named parameters, to avoid SQL injection
- [x] If you are changing dependencies, check the Snyk status check or the dashboard to ensure you are not introducing new high/critical vulnerabilities
- [x] Assume that inputs to the API can be malicious, and sanitize and/or check for Denial of Service type values, e.g., massive sizes
- [x] Do not serve user-uploaded binary images through the Dockstore API
- [x] Ensure that endpoints that only allow privileged access enforce that with the `@RolesAllowed` annotation
- [x] Do not create cookies, although this may change in the future
- [x] If this PR is for a user-facing feature, create and link a documentation ticket for this feature (usually in the same milestone as the linked issue). Style points if you create a documentation PR directly and link that instead.

### CODE REVIEW

**Comment by denis-yuen** [comment]
Mixed feelings about this, so going to break this down

1) This does seem to install cwltool properly in Ubuntu 20.04 (tested via Docker container). This has the older Python version i.e. this fix seems backwards compatible although we have to rely on integration tests to see that the CLI integration works
2) This doesn't work in Ubuntu 24.04, which has the newer python version. This Ubuntu version was released in April. 
3) That said, the reason it doesn't work is that the newer version of Ubuntu essentially requires `venv ` and we don't mention this in our quick start (and we haven't tested it either) https://stackoverflow.com/questions/75602063/pip-install-r-requirements-txt-is-failing-this-environment-is-externally-mana/76084249#76084249 
4) The Python install/support ecosystem is terrible but this is known https://xkcd.com/1987/
5) User was on Mac

So my superficial thought was to warn users away from Python 3.12 as a hotfix and upgrade cwltool to a new version in the 1.16 webservice. But breaking it down, it kinda seems like Ubuntu 22.04 users are kinda screwed either way and this (hopefully) wouldn't affect Ubuntu 22.04 users (if our tests are ok). 

I cannot speak to the Mac experience. 

To putting that all together, I think I'm tentatively ok with this if 
a) this works on Mac in a backwards compatible manner
b) this new requirements file/webservice is tested in a CLI pr as well to ensure that this isn't a big breaking change 

e.g. review the oicr wiki release and release something like `1.15.4-prebuild.0` and test the CLI with it

File: dockstore-common/src/main/java/io/dockstore/common/PipHelper.java
Comment (denis-yuen) [inline dockstore-common/src/main/java/io/dockstore/common/PipHelper.java:38]: Pending CLI pr, I think this is the CLI version and we don't know yet if this is backwards compatible

**Review by denis-yuen — COMMENTED** [review]
see comment above

**Comment by david4096** [comment]
Notes:

1) I think covering 3 ubuntu versions is confusing things a little. 20.04 uses python 3.8 and I cannot use these dependencies to get it working there. Could you share what you did?
2, 3) 24.04 does suggest using virtualenvironment and there is a useful error when you go to install it. Seems like out of the scope of this ticket but probably should be documented. `cwltool` is actually in the ubuntu repositories so I think the best way to install going forward for linux in `apt install cwltool`

I tested on the ubuntu docker too.

Understood that we need to update the CLI to test this end-to-end. As a matter of process, should that be blocking here or we merge this first?

**Comment by denis-yuen** [comment]
> 1. `cwltool` is actually in the ubuntu repositories so I think the best way to install going forward for linux in `apt install cwltool`

Oooh, nice. That is cleaner

**Comment by denis-yuen** [comment]
> 1. 20.04 uses python 3.8 and I cannot use these dependencies to get it working there. Could you share what you did?

Ack, I think I had a typo in there. I tested 22.04 and 24.04. 20.04 fails for me as well.

File: dockstore-common/src/main/java/io/dockstore/common/PipHelper.java
Comment (coverbeck) [reply dockstore-common/src/main/java/io/dockstore/common/PipHelper.java:38]: I don't follow, but there's a good chance I'm missing something... This is the CLI version, only the 1.15 CLI will hit this. Shouldn't this just work?

**Review by coverbeck — COMMENTED** [review]
I'm a little out of my depth on the Python stuff. I feel like this works, but we should also have additional doc for `venv`; I know it can get confusing when you're just beginning and following the getting started instructions doesn't work... I know I would have been thrown for a loop when I started Dockstore.

I'm just going to Comment through instead of Approve so I can get another look at things after addressing (or not) Denis' feedback

File: dockstore-common/src/main/java/io/dockstore/common/PipHelper.java
Comment (denis-yuen) [reply dockstore-common/src/main/java/io/dockstore/common/PipHelper.java:38]: Elaborating, we've released a 1.15 cli. After this change, existing users with a 1.15 cli will get the new requirements instead of the old 1.14.0 requirements from our webservice. 

At the time the comment was written, I didn't know if the new dependencies are backwards compatible. 
Afterwards, we found out that they actually aren't backwards compatible for one reason,. they will fail on Ubuntu 20.04.  There may be other reasons that our 1.15.0 CLI code is not compatible with these requirements, we need to run cli tests with the proposed webservice to find  out.

**Comment by mr-c** [comment]
> Mixed feelings about this, so going to break this down
> 
>     1. This does seem to install cwltool properly in Ubuntu 20.04 (tested via Docker container). This has the older Python version i.e. this fix seems backwards compatible although we have to rely on integration tests to see that the CLI integration works
> 
>     2. This doesn't work in Ubuntu 24.04, which has the newer python version. This Ubuntu version was released in April.
> 
>     3. That said, the reason it doesn't work is that the newer version of Ubuntu essentially requires `venv ` and we don't mention this in our quick start (and we haven't tested it either) https://stackoverflow.com/questions/75602063/pip-install-r-requirements-txt-is-failing-this-environment-is-externally-mana/76084249#76084249
> 
>     4. The Python install/support ecosystem is terrible but this is known https://xkcd.com/1987/
> 
>     5. User was on Mac
> 
> 
> So my superficial thought was to warn users away from Python 3.12 as a hotfix and upgrade cwltool to a new version in the 1.16 webservice. But breaking it down, it kinda seems like Ubuntu 22.04 users are kinda screwed either way and this (hopefully) wouldn't affect Ubuntu 22.04 users (if our tests are ok).
> 
> I cannot speak to the Mac experience.
> 
> To putting that all together, I think I'm tentatively ok with this if a) this works on Mac in a backwards compatible manner b) this new requirements file/webservice is tested in a CLI pr as well to ensure that this isn't a big breaking change c) we spin off a ticket to add further warnings for Ubuntu 22.04 users (not really a fault of this ticket, just that Ubuntu 22.04 is the one with Python 3.12.x)
> 
> e.g. review the oicr wiki release and release something like `1.15.4-prebuild.0` and test the CLI with it

I recommend that you update your docs to use `pipx`, it works on Python 3.8+

**Comment by mr-c** [comment]
> > 1. `cwltool` is actually in the ubuntu repositories so I think the best way to install going forward for linux in `apt install cwltool`
> 
> Oooh, nice. That is cleaner It also avoids that [PEP 668](https://peps.python.org/pep-0668/) issue, but we do kinda lose the benefit of pinning a version we've tested with. But I think the benefits outweight the drawbacks, at least for 24.04 For 22.04 it installs `3.1.20220224085855` which is significantly older than the `3.1.20230201224320` we tested with

Hello, yes Ubuntu ships a snapshot of the `cwltool` package from Debian, for which I'm also the maintainer.  If you really want to use the Ubuntu packages, then I suggest requesting them to backport a newer version of `cwltool` (and/or assisting them to do so): https://wiki.ubuntu.com/UbuntuBackports

**Review by svonworl — COMMENTED** [review]
Comments more focused towards the long term:
Seems like someone who has enough expertise to run the dockstore cli could follow the cwltool installation instructions ( https://cwltool.readthedocs.io/en/latest/#install,), so could the CLI simply detect a stale or missing cwltool and a link to the installation instructions in the error message?  Is there a reason why we need to provide the `requirements.txt`?

**Comment by david4096** [comment]
Thanks @mr-c we will update our docs to use pipx instead.

I made a tagged release of this branch so we could test the CLI against it. https://github.com/dockstore/dockstore/tree/1.15.4-alpha.1

Continued work on this is going https://github.com/dockstore/dockstore-cli/pull/291

**Comment by david4096** [comment]
The CLI PR that is tested against the tagged release of this branch passes, so this should be able to be merged. Following that I will make another tagged release of the webservice and point the CLI PR to it before merging.

https://github.com/dockstore/dockstore-cli/pull/291

**Comment by david4096** [comment]
> Comments more focused towards the long term: Seems like someone who has enough expertise to run the dockstore cli could follow the cwltool installation instructions ( https://cwltool.readthedocs.io/en/latest/#install,), so could the CLI simply detect a stale or missing cwltool and a link to the installation instructions in the error message? Is there a reason why we need to provide the `requirements.txt`?

That's a good point. The intention is to give good reproducibility guarantees as in, for a given version of Dockstore we did test cwltool and the given set of dependencies. The nice-to-have of the dockstore-cli is that they don't have to know how to install cwltool, but you're probably right that in most cases they could. Installing something for java is often easier than python though

**Review by denis-yuen — COMMENTED** [review]
@david4096  A little unclear understandably since the changes go across repos. 

Just to make sure I understand the proposal:
Is the idea to
a) CLI 1.14 users will be unchanged and will likely be incompatible with Python 2.12
b) CLI 1.15 users will get the new requirements file across the board (means Ubuntu 20.04 may fail)
c) GUI instructions will recommend `pipx` for ubuntu
d) some builds will likely remain using  requirements files (as opposed to  a pinned apt version) but switch from `pip3` to `pipx`
?

**Comment by david4096** [comment]
> @david4096 A little unclear understandably since the changes go across repos.
> 
> Just to make sure I understand the proposal: Is the idea to a) CLI 1.14 users will be unchanged and will likely be incompatible with Python 2.12 

Yes, the 1.14 requirements stayed the same and people with 3.12 will not have to install from the more recent requirements.

b) CLI 1.15 users will get the new requirements file across the board (means Ubuntu 20.04 may fail) 

Yes, the oldest ubuntu (which is still covered for security but not feature updates btw) will be left behind.

c) GUI instructions will recommend `pipx` for ubuntu 

Thanks for the reminder on this, I went to look in dockstore-documentation but this is actually in the ui. I'll put in a PR for this too

d) some builds will likely remain using requirements files (as opposed to a pinned `apt` version) but switch from `pip3` to `pipx` ?

I think for some builds you mean "our builds" where we control the python version, so we can use pip3 (the current builds all pass using this method)

**Review by coverbeck — APPROVED** [review]

**Review by svonworl — APPROVED** [review]

**Comment by denis-yuen** [comment]
> I think for some builds you mean "our builds" where we control the python version, so we can use pip3 (the current builds all pass using this method)

Yup and :+1:

**Review by denis-yuen — APPROVED** [review]

**Comment by david4096** [comment]
https://github.com/dockstore/playbooks/wiki/Updating-cwltool-version

**Pull request merged**
